Add parallel-rustc blog post. (#1164)
rust-lang/blog.rust-lang.org: posts/inside-rust/2023-11-10-parallel-rustc.md

---
layout: post
title: Faster compilation with the parallel front-end in nightly
author: Nicholas Nethercote
team: The Parallel Rustc Working Group <https://www.rust-lang.org/governance/teams/compiler#Parallel%20rustc%20working%20group>
---

The Rust compiler's front-end can now use parallel execution to significantly
reduce compile times. To try it out, run the nightly compiler with the `-Z

> threads=8` command line option.

Reviewer: Maybe mention that it's more experimental with known bugs immediately for TLDR?

Again, there are several things worth nothing.
- Front-end execution takes 5.9 seconds (down from 10.2 seconds).
- Back-end execution takes 5.3 seconds (down from 6.2 seconds), and the LLVM
  threads are running for 4.9 seconds of that (down from 5.9 seconds).
- There are seven additional threads labelled "rustc" operating in the
  front-end. The reduced front-end time shows they are reasonably effective,
  but the thread utilization is patchy, with the eight threads all having
  periods of inactivity. There is room for significant improvement here.
- Eight of the LLVM threads start at the same time. This is because the eight
  "rustc" threads create the LLVM IR for eight codegen units in parallel. (For
  seven of those threads that is the only work they do in the back-end.) After
  that, the staircase effect returns because only one "rustc" thread does LLVM
  IR generation while seven or more other LLVM threads are active. If the
  number of threads used by the front-end was changed to 16 the staircase shape
  would disappear entirely, though in this case the final execution time would
  barely change.

### Putting it all together

> Rust compilation has long benefited from intraprocess parallelism, via Cargo,

Reviewer 2: ```suggestion
Rust compilation has long benefited from interprocess parallelism, via Cargo,
```

> and from interprocess parallelism in the back-end. It can now also benefit from

Reviewer 2: ```suggestion
and from intraprocess parallelism in the back-end. It can now also benefit from
```

> interprocess parallelism in the front-end.

Reviewer 2: ```suggestion
intraprocess parallelism in the front-end.
```

> Attentive readers might be wondering how intraprocess parallelism and
> interprocess parallelism interact. If we have 20 parallel rustc invocations and
> each one can have up to 16 threads running, might we end up 100s of threads on
> a machine with only 10s of cores, resulting in inefficient execution as the OS
> tries its best to schedule them?

Reviewer 3: I first read this as "one hundred seconds" thinking in CPU time, before realizing it meant "hundreds". This might be more clear:

```suggestion
Attentive readers might be wondering how intraprocess parallelism and
interprocess parallelism interact. If we have 20 parallel rustc invocations and
each one can have up to 16 threads running, might we end up with hundreds of threads on
a machine with only tens of cores, resulting in inefficient execution as the OS
tries its best to schedule them?
```

Fortunately no. The compiler uses the [jobserver
protocol](https://www.gnu.org/software/make/manual/html_node/POSIX-Jobserver.html)
to limit the number of threads it creates. If a lot of interprocess parallelism
is occuring, intraprocess parallelism will be limited appropriately, and
the number of threads will not exceed the number of cores.

## How to use it

As of (XXX: date enabled) the nightly compiler is [shipped with the parallel
front-end enabled](https://github.com/rust-lang/rust/pull/117435). However,
**by default it runs in single-threaded mode** and won't reduce compile times.

> Keen users who want to try multi-threaded mode should use the `-Z threads=8`

Reviewer 4: maybe mention or link to somewhere that mentions the ways to set rustc flags with cargo like RUSTFLAGS or config.toml?

> Keen users who want to try multi-threaded mode should use the `-Z threads=8`

Reviewer: I might mention that `-Z threads=n` where `n` is above the physical core count would not be recommended.

Reviewer 5: Maybe not. On my 48-core machine, the performance of 16 threads is not even as good as 8 threads. And it can be observed that the utilization rate of many threads is low.

Author: Based on these comments, I think the current text is ok.

option.

This default may be surprising. Why parallelize the front-end and then run it
in single-threaded mode? The answer is simple: caution. This is a big change!
The parallel front-end has a lot of new code. Single-threaded mode exercises
most of the new code, but excludes the possibility of threading bugs such as
deadlocks that still occasionally affect multi-threaded mode. Parallel
execution is harder to get right than serial execution, even in Rust. For this

> reason the parallel front-end also won't be shipped in Beta or Stable releases

Reviewer 6: This is quite crucial, because some people might just read the headline and think that the parallel compiler will be in stable in a few weeks. I'd put it in bold, and perhaps also move/copy it to the very beginning of the post.

for some time.

### Performance effects

When the parallel front-end is run in single-threaded mode, compilation times

> are typically within 2% of the serial front-end, which should be barely

Reviewer 6: It's not clear from this that the performance can be worse, I think.

noticeable.

When the parallel front-end is run in multi-threaded mode with `-Z threads=8`,

> out [measurements on real-world

Reviewer 7: ```suggestion
our [measurements on real-world
```

> code](https://github.com/rust-lang/compiler-team/issues/681) on show that

Reviewer 7: ```suggestion
code](https://github.com/rust-lang/compiler-team/issues/681) show that
```

compile times can be reduced by up to 50%, though the affects vary widely and
depend greatly on the characteristics of the code being compiled and the build
configuration. For example, dev builds are likely to see bigger improvements
than release builds because release builds usually spend more time doing
optimizations in the back-end. A small number of cases compile more slowly in

> multi-threaded mode than single-threaded mode.

Reviewer 5: I think we can mention that most of the cases that get regressions are those crates take small number of time to compile. For example, `hello world` and codes under `incr-unchanged` scenarios.

We recommend eight threads because this is the configuration we have tested the
most and it is known to give good results. Values lower than eight will see
smaller benefits. Values greater than eight will give diminishing returns and
may even give worse performance.

If a 50% improvement seems low when going from one to eight threads, recall
from the explanation above that the front-end only accounts for part of compile
times, and the back-end is already parallel. You can't beat [Amdahl's
Law](https://en.wikipedia.org/wiki/Amdahl%27s_law).

Memory usage can increase significantly in multi-threaded mode. This is
unsurprising given that various tasks, each of which requires a certain amount
of memory, are now executing in parallel. We have seen increases of up to 35%.

### Correctness

Reliability in single-threaded mode should be high.

In multi-threaded mode there are some known bugs, including deadlocks. If
compilation hangs, you have probably hit one of them.

### Feedback

> If you have any problems with the parallel front-end, please [file an issue

Reviewer: Currently panics with `no ImplicitCtxt stored in tls` and `deadlock detected` are known (no real deadlocks). Perhaps we should explicitly mention these and encourage reporting if users find other bugs?

Author: I see https://github.com/rust-lang/rust/issues/111522 for the `ImplicitCtxt` one. Do we have an open issue for `deadlock detected`?

Reviewer: https://github.com/rust-lang/rust/issues/115223 and https://github.com/rust-lang/rust/issues/111521. I renamed the panic in https://github.com/rust-lang/rust/pull/115668. @matthiaskrgr Could you rename the issues so we don't gain extra ones?

Reviewer 8: Hi, sorry, rename what to what?

Author: I've tweaked the feedback section to say "If your problem does not match any of the existing issues, please file a new issue." Is that good enough?

Reviewer: That's probably fine.

marked with the "WG-compiler-parallel"
label](https://github.com/rust-lang/rust/labels/WG-compiler-parallel).
That link also shows existing known problems.

For more general feedback, please start a discussion on the [wg-parallel-rustc
Zulip
channel](https://rust-lang.zulipchat.com/#narrow/stream/187679-t-compiler.2Fwg-parallel-rustc).
We are particularly interested to hear the performance effects on the code you
care about.

# Future work

We are working to improve the performance of the parallel front-end. As the
graphs above showed, there is room to improve the utilization of the threads in
the front-end.

We are also working to iron out the remaining bugs in multi-threaded mode. We
aim to stabilize the `-Z threads` option and ship the parallel front-end in

> multi-threaded mode on Stable releases in 2024.

Reviewer 7: Are we aiming to make n > 1 the default in 2024, or just n=1 the default on stable?

Author: n > 1, as far as I know. (That's what "multi-threaded mode" refers to.) That's the whole point :)

Reviewer 7: Ah, I see. This makes sense.

has continually improved compiler performance for several years. For example,
in the first 10 months of 2023, there were mean reductions in compile time of
[13%](https://perf.rust-lang.org/compare.html?start=2023-01-01&end=2023-10-31&stat=wall-time&nonRelevant=true),
in peak memory use of
[15%](https://perf.rust-lang.org/compare.html?start=2023-01-01&end=2023-10-31&stat=max-rss&nonRelevant=true),
and in binary size of
[7%](https://perf.rust-lang.org/compare.html?start=2023-01-01&end=2023-10-31&stat=size%3Alinked_artifact&nonRelevant=true),
as measured by our performance suite.

However, at this point the compiler has been heavily optimized and new
improvements are hard to find. There is no low-hanging fruit remaining. 

But there is one piece of large but high-hanging fruit: parallelism. Current
Rust compiler users benefit from two kinds of parallelism, and the newly
parallel front-end adds a third kind.

### Existing interprocess parallelism

When you compile a Rust program, Cargo launches multiple rustc processes,
compiling multiple crates in parallel. This works well. Try compiling a large
Rust program with the `-j1` flag to disable this parallelization and it will

> take a lot longer than normal.

Reviewer 9: do we need to say here that `-j1` would not just "disable this parallelization" but would also disable the parallelism in rustc's backend?

Author: Is that true? Either way, `-j1` will be slower and I think that's an unimportant detail.

Reviewer: Maybe `to disable parallelization`?

Reviewer 9: I agree it's not particularly important because the point is that we do have parallelism already and it speeds things up considerably, but it is true yes. (Building a leaf crate at j1 will show the difference, IIRC `syn` is particularly helped by rustc's intraprocess parallelism today.)

Reviewer 9: @Mark-Simulacrum Happy to have this resolved by you or @nnethercote as the distinction is not absolutely needed, but can't mark it resolved myself.
